plugins/nextcloud/index.php

```diff
@@ -36,26 +56,80 @@ public function Init() : void
 			$this->addTemplate('templates/PopupsNextcloudFiles.html');
 			$this->addTemplate('templates/PopupsNextcloudCalendars.html');
 
-//			$this->addHook('login.credentials.step-2', 'loginCredentials2');
-//			$this->addHook('login.credentials', 'loginCredentials');
+			//			$this->addHook('login.credentials.step-2', 'loginCredentials2');
+			//			$this->addHook('login.credentials', 'loginCredentials');
 			$this->addHook('imap.before-login', 'beforeLogin');
 			$this->addHook('smtp.before-login', 'beforeLogin');
 			$this->addHook('sieve.before-login', 'beforeLogin');
+
+			if ($this->Config()->Get('plugin', self::ENABLE_NC_ADDRESSBOOK_KEY, self::ENABLE_NC_ADDRESSBOOK_DEFAULT_VALUE)) {
+				$this->addJs('js/addressbook.js');
+				$this->addJsonHook('NextcloudGetAddressBooks', 'GetAddressBooks');
+				$this->addJsonHook('NextcloudUpdateAddressBook', 'UpdateAddressBook');
+			}
+
+			if ($this->Config()->Get('plugin', self::DISABLE_INHOUSE_ADDRESSBOOK_KEY, self::DISABLE_INHOUSE_ADDRESSBOOK_DEFAULT_VALUE)) {
+				$this->addJs('js/hideInhouseAddressbook.js');
+			}
 		} else {
 			\SnappyMail\Log::debug('Nextcloud', 'NOT integrated');
 			// \OC::$server->getConfig()->getAppValue('snappymail', 'snappymail-no-embed');
 			$this->addHook('main.content-security-policy', 'ContentSecurityPolicy');
 		}
 	}
 
+	public function GetAddressBooks()
+	{
+		$addressBooks = [];
+
+		$contactsManager = \OC::$server->getContactsManager();
+
+		$defaultUri = $this->Config()->Get('plugin', self::DEFAULT_ADDRESSBOOK_URI_KEY, self::DEFAULT_ADDRESSBOOK_URI);
+		$selectedUri = $this->UserSettings()->GetConf(self::ADDRESSBOOK_SETTINGS_KEY, $defaultUri);
+		$ignoreSystemAddressbook = $this->Config()->Get('plugin', self::IGNORE_SYSTEM_ADDRESSBOOK_KEY, self::IGNORE_SYSTEM_ADDRESSBOOK_DEFAULT_VALUE);
+
+		foreach ($contactsManager->getUserAddressBooks() as $addressBook) {
+			if ($ignoreSystemAddressbook && $addressBook->isSystemAddressBook()) {
+				$contactsManager->unregisterAddressBook($addressBook);
+				continue;
+			}
+
+			$book = new AddressBook();
+			$book->uri = $addressBook->getUri();
+			$book->name = $addressBook->getDisplayName();
+			if (strcmp($selectedUri, $book->uri) === 0) {
+				$book->selected = true;
+			}
+
+			$addressBooks[] = $book;
+		}
+
+
+		return $this->jsonResponse(__FUNCTION__, array(
+			'addressbooks' => json_encode($addressBooks)
+		));
+	}
+
+	public function UpdateAddressBook(): array
+	{
+		$uri = $this->jsonParam('uri');
+		$oSettings = $this->UserSettings();
+		if (\is_string($uri)) {
+			$oSettings->SetConf(self::ADDRESSBOOK_SETTINGS_KEY, $uri);
+			$this->SettingsProvider()->Save($this->Account(), $oSettings);
+		}
+		return $this->jsonResponse(__FUNCTION__, true);
+	}
+
+
 	public function ContentSecurityPolicy(\SnappyMail\HTTP\CSP $CSP)
 	{
 		if (\method_exists($CSP, 'add')) {
 			$CSP->add('frame-ancestors', "'self'");
 		}
 	}
 
-	public function Supported() : string
```

Comment:
Can we avoid all these lint changes @fahim44 ?

The snappymail project might be following a different lint strategy than what you have locally

Reply:
ok will try...

Reply:
@akhil1508 please check again